### PR TITLE
create multiview rtt

### PR DESCRIPTION
**Pull request by TrevorDev**
https://github.com/BabylonJS/Babylon.js/issues/4989
TestPG: http://localhost:1338/Playground/index-local.html#LVL53U

### CODE REVIEW

File: src/scene.ts [hunk 1]
```diff
@@ -2546,16 +2548,23 @@ export class Scene extends AbstractScene implements IAnimatable {
      * @param projection defines the Projection matrix to use
      */
     public setTransformMatrix(view: Matrix, projection: Matrix): void {
-        if (this._viewUpdateFlag === view.updateFlag && this._projectionUpdateFlag === projection.updateFlag) {
+        this._setMultiviewTransformMatrix(view, projection);
```
Comment (deltakosh) [inline src/scene.ts]: why not adding the additional parameter to setTransformmatrix? The _setMultiviewTransformmatrix is a bit misleading as it also takes care of non multiview version

File: src/Materials/Textures/renderTargetTexture.ts [hunk 2]
```diff
@@ -975,8 +985,76 @@ export class RenderTargetTexture extends Texture {
...
 
 Texture._CreateRenderTargetTexture = (name: string, renderTargetSize: number, scene: Scene, generateMipMaps: boolean) => {
     return new RenderTargetTexture(name, renderTargetSize, scene, generateMipMaps);
 };
+
+/**
+ * Renders to multiple views with a single draw call
+ * @see https://www.khronos.org/registry/webgl/extensions/WEBGL_multiview/
+ */
+export class MultiviewRenderTarget extends RenderTargetTexture {
```
[the remark this answers is not in the document]
Comment (TrevorDev) [reply src/Materials/Textures/renderTargetTexture.ts]: It is all one word according to the spec https://www.khronos.org/registry/webgl/extensions/WEBGL_multiview/

File: src/scene.ts [hunk 3]
```diff
@@ -1588,6 +1589,7 @@ export class Scene extends AbstractScene implements IAnimatable {
     private _createUbo(): void {
         this._sceneUbo = new UniformBuffer(this._engine, undefined, true);
         this._sceneUbo.addUniform("viewProjection", 16);
+        this._sceneUbo.addUniform("viewProjectionR", 16);
```
[the remark this answers is not in the document]
Comment (TrevorDev) [reply src/scene.ts]: This is only created once in the constructor, is it ok to leave this? Or should it be recreated whenever multiview is set?

**Comment by TrevorDev** [comment]
Playground I am using for testing: https://playground.babylonjs.com/#FTJDIR#11 

Right now it only is able to be specified as an output texture for a a camera with child cameras (This will work really well for webXR). I have it working with a custom shader to present the image so I know it is drawing and is readable but am still trying to figure out how to replace the alternateCamera logic with this.

File: src/scene.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (deltakosh) [reply src/scene.ts]: it is created once but if multiview is not actie then we have unused data in the ubo. I will recommend not using the ubo for the multiview at all

**Comment by deltakosh** [comment]
ok no rush

File: src/Materials/Background/backgroundMaterial.ts [hunk 4]
```diff
@@ -675,6 +676,12 @@ export class BackgroundMaterial extends PushMaterial {
         MaterialHelper.PrepareDefinesForLights(scene, mesh, defines, false, this._maxSimultaneousLights);
         defines._needNormals = true;
 
+        // Multiview
+        if (scene.activeCamera) {
+            defines.MULTIVIEW = (scene.activeCamera.outputRenderTarget !== null && scene.activeCamera.outputRenderTarget.getViewCount() > 1);
+            defines.markAsUnprocessed();
```
Comment (deltakosh) [inline src/Materials/Background/backgroundMaterial.ts]: this should be only if it was not the case before (no need to recompile then)

File: src/Materials/PBR/pbrBaseMaterial.ts [hunk 5]
```diff
@@ -1236,6 +1241,12 @@ export abstract class PBRBaseMaterial extends PushMaterial {
         MaterialHelper.PrepareDefinesForLights(scene, mesh, defines, true, this._maxSimultaneousLights, this._disableLighting);
         defines._needNormals = true;
 
+        // Multiview
+        if (scene.activeCamera) {
+            defines.MULTIVIEW = (scene.activeCamera.outputRenderTarget !== null && scene.activeCamera.outputRenderTarget.getViewCount() > 1);
+            defines.markAsUnprocessed();
```
Comment (deltakosh) [inline src/Materials/PBR/pbrBaseMaterial.ts]: same

File: src/Materials/standardMaterial.ts [hunk 6]
```diff
@@ -792,6 +793,13 @@ export class StandardMaterial extends PushMaterial {
 
         // Lights
         defines._needNormals = MaterialHelper.PrepareDefinesForLights(scene, mesh, defines, true, this._maxSimultaneousLights, this._disableLighting);
+        if (scene.activeCamera) {
+            var previousMultiview = defines.MULTIVIEW;
+            defines.MULTIVIEW = (scene.activeCamera.outputRenderTarget !== null && scene.activeCamera.outputRenderTarget.getViewCount() > 1);
+            if (defines.MULTIVIEW != previousMultiview) {
```
Comment (deltakosh) [inline src/Materials/standardMaterial.ts]: yeah..like that :D

File: src/PostProcesses/vrDistortionCorrectionPostProcess.ts [hunk 7]
```diff
@@ -53,3 +53,30 @@ export class VRDistortionCorrectionPostProcess extends PostProcess {
...
+     * @param camera camera to be applied to
+     * @param scaleFactor scaling factor to the size of the output texture
+     */
+    constructor(name: string, camera: Camera, scaleFactor: number) {
+        super(name, "vrMultiviewToSingleview", ["imageIndex"], ["multiviewSampler"], scaleFactor, camera, Texture.BILINEAR_SAMPLINGMODE);
+
+        this.onSizeChangedObservable.add(() => {
+        });
+        this.onApplyObservable.add((effect: Effect) => {
+            if (camera._scene.activeCamera && camera._scene.activeCamera.name == "_L") {
```
Comment (deltakosh) [inline src/PostProcesses/vrDistortionCorrectionPostProcess.ts]: well can we be a little bit dependent on something easy to change? Not a big fan of testing a string here

File: src/scene.ts [hunk 8]
```diff
@@ -4120,27 +4092,53 @@ export class Scene extends AbstractScene implements IAnimatable {
...
+
+            // Consume the multiview texture through a shader for each eye
+            for (var index = 0; index < camera._rigCameras.length; index++) {
+                var engine = this._engine;
+                this._activeCamera = camera._rigCameras[index];
+                engine.setViewport(this._activeCamera.viewport);
+                if (this.postProcessManager) {
+                    this.postProcessManager._prepareFrame();
+                }
+                if (this.postProcessManager) {
```
Comment (deltakosh) [inline src/scene.ts]: could be in previous if

**Comment by TrevorDev** [comment]
When working on samples with webVR, I ran into a couple issues, one thing I'm still trying to figure out is that the environment.ground mesh is not visible when using webVR + multiview . If i set opacity fresnel to false in the inspector it works but I'm still looking into how to fix that. Its weird because multiview + vrDeviceCamera works.

File: src/Cameras/VR/webVRCamera.ts [hunk 9]
```diff
@@ -140,6 +142,10 @@ export interface WebVROptions {
      */
     defaultHeight?: number;
 
+    /**
+     * If multiview should be used if availible (default: false)
```
Comment (TrevorDev) [inline src/Cameras/VR/webVRCamera.ts]: One thing to note is multiview is disabled by default and needs to be enabled as existing apps might expect that the output texture of the camera will not be a texture array

Comment (deltakosh) [reply src/Cameras/VR/webVRCamera.ts]: Can you provide some code examples on how to use it?

Comment (deltakosh) [reply src/Cameras/VR/webVRCamera.ts]: Well actually it will be in the doc so no worries for now

**Comment by deltakosh** [comment]
ok besides the bug you identified I think the PR is pretty clean and ready to be merged
Also please report the perf boost in each individual test cases

**Comment by TrevorDev** [comment]
Seeing between 1.5x - 2.0x speedup

**WebVR Helper**
A lot of spheres + pbr + background material: https://playground.babylonjs.com/#3DN3RJ
With muliview:
45fps
Without:
24fps

HillValley: https://playground.babylonjs.com/#CZJYG7#1
With muliview:
image
Without:
image



**VRDeviceOrientationFreeCamera**
A lot of spheres + pbr + background material: https://playground.babylonjs.com#4YK425
With muliview:
55fps
Without:
28fps

HillValley: https://playground.babylonjs.com#CZJYG7
With muliview:
image
Without:
image

**Pull request merged**